### PR TITLE
tests: add check_output test helper

### PR DESCRIPTION
**Pull request by bartlomieju**
Towards https://github.com/boa-dev/boa/issues/803

### CODE REVIEW

File: boa/src/value/tests.rs [hunk 1]
```diff
@@ -64,62 +64,47 @@ fn number_is_true() {
...
+            ("'-0' == 0", "true"),
+            ("0 == NaN", "false"),
+            ("'foo' == NaN", "false"),
+            ("NaN == NaN", "false"),
+            (
+                "Number.POSITIVE_INFINITY === Number.POSITIVE_INFINITY",
+                "true",
+            ),
+            (
+                "Number.NEGAVIVE_INFINITY === Number.NEGAVIVE_INFINITY",
```
Comment (jedel1043) [inline boa/src/value/tests.rs:104]:
```suggestion
                "Number.NEGATIVE_INFINITY === Number.NEGATIVE_INFINITY",
```
Realized there's a typo here 😅, this is evaluating `undefined === undefined`.

**Comment by jedel1043** [comment]
Maybe it would be good to allow the execution of intermediate instructions between tests. That way we could cleanup some tests like:
```Rust
let mut context = Context::new();
let init = r#"
    var iterator = [1, 2, 3].keys();
    var next = iterator.next();
"#;
forward(&mut context, init);
assert_eq!(forward(&mut context, "next.value"), "0");
assert_eq!(forward(&mut context, "next.done"), "false");
forward(&mut context, "next = iterator.next()");
assert_eq!(forward(&mut context, "next.value"), "1");
assert_eq!(forward(&mut context, "next.done"), "false");
```

I'd suggest creating a `TestAction` enum with the variants `TestEq((&str, &str))` and `Execute(&str)`. This would change the previous test to:

```Rust
let init = r#"
    var iterator = [1, 2, 3].keys();
    var next = iterator.next();
"#;
check_output(
    &[
        TestAction::Execute(init),
        TestAction::TestEq(("next.value", "0")),
        TestAction::TestEq(("next.done", "false")),
        TestAction::Execute("next = iterator.next()"),
        TestAction::TestEq(("next.value", "1")),
        TestAction::TestEq(("next.done", "false")),
    ]
)
```
Notice how we don't need the `maybe_init` parameter, we just pass it as a statement to be executed.

**Comment by bartlomieju** [comment]
Sure, I can rework it to your suggestion

**Comment by bartlomieju** [comment]
@jedel1043 I applied your suggestion, please let me know if this is satisfactory

**Review by HalidOdat — APPROVED** [review]
This looks really good! We can merge this as is and refactor more in future PRs

File: boa/src/lib.rs [hunk 2]
```diff
@@ -153,25 +153,43 @@ pub(crate) fn exec<T: AsRef<[u8]>>(src: T) -> String {
...
+            }
+            TestAction::TestStartsWith(case, expected) => {
+                assert!(
+                    &forward(&mut context, case).starts_with(expected),
+                    "Test case {} ('{}')",
+                    i + 1,
+                    case
+                );
+            }
+        }
```
Comment (jedel1043) [inline boa/src/lib.rs:200]:
```suggestion
    let mut i = 1;
    for action in actions {
        match action {
            TestAction::Execute(src) => {
                forward(&mut context, src);
            }
            TestAction::TestEq(case, expected) => {
                assert_eq!(
                    &forward(&mut context, case),
                    expected,
                    "Test case {} ('{}')",
                    i,
                    case
                );
                i += 1;
            }
            TestAction::TestStartsWith(case, expected) => {
                assert!(
                    &forward(&mut context, case).starts_with(expected),
                    "Test case {} ('{}')",
                    i,
                    case
                );
                i += 1;
            }
        }
```

Let's skip counting between executions

**Review by jedel1043 — CHANGES_REQUESTED** [review]
Just a small nitpick about the test counter. Everything else is pretty good! We should be able to extend `TestAction` in case we find more common use cases.

File: boa/src/lib.rs (hunk 2 above)
Comment (bartlomieju) [reply boa/src/lib.rs]: Applied

**Review by jedel1043 — APPROVED** [review]
Looks great to me!

**Pull request merged**